Marketplace/src/main/java/com/marketplace/service/category/JdbcCategoryService.java

```diff
@@ -0,0 +1,145 @@
+package com.marketplace.service.category;
+
+import com.marketplace.config.ImageLoader;
+import com.marketplace.repository.category.CategoryRepository;
+import com.marketplace.repository.category.DbCategory;
+import com.marketplace.repository.category.DbCharacteristic;
+import com.marketplace.repository.category.ProductQuery;
+import com.marketplace.config.exception.NonExistingEntityException;
+import com.marketplace.config.exception.ParentCategoryException;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class JdbcCategoryService implements CategoryService {
+
+    private final CategoryRepository categoryRepository;
+    private final ImageLoader imageLoader;
+
+    public JdbcCategoryService(CategoryRepository categoryRepository, ImageLoader imageLoader) {
+        this.categoryRepository = categoryRepository;
+        this.imageLoader = imageLoader;
+    }
+
+    private Map<String, Set<DbCharacteristic>> getGroupedByName(long categoryId) {
+        return categoryRepository.getCharacteristics(categoryId)
+                .stream()
+                .collect(
+                        Collectors.groupingBy(
+                                DbCharacteristic::getName,
+                                Collectors.toSet()
+                        )
+                );
+    }
+
+    private Characteristic combineCharacteristicValues(Set<DbCharacteristic> set) {
+        Characteristic result = new Characteristic();
+        for (DbCharacteristic c : set) {
+            result.getValues().put(
+                    c.getCharacteristicId(),
+                    c.getCharacteristicValue()
+            );
+        }
+        return result;
+    }
+
+    private List<Characteristic> combineCharacteristicValues(Map<String, Set<DbCharacteristic>> map) {
+        return map.entrySet()
+                .stream()
+                .map(entry -> {
+                    return combineCharacteristicValues(entry.getValue());
+                })
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public boolean isParentCategory(long categoryId) {
+        return categoryRepository.isParentCategory(categoryId);
+    }
+
+    @Override
+    public long addCategory(Category category) throws IOException {
+
+        if (!categoryRepository.categoryExists(category.getParentId())) {
+            throw new NonExistingEntityException(
+                    "parent category for this category does not exist");
+        }
+
+        String img_location = imageLoader.save(
```

Comment:
Only camel case in Java code. https://google.github.io/styleguide/javaguide.html#s5.2.7-local-variable-names